File: 백준/Silver/24371. БРОЙ ПОДНИЗОВЕ/БРОЙ ПОДНИЗОВЕ.cc
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <map>
#include <set>

using namespace std;

int main() {
    string S, P;
    cin >> S >> P;

    bool is_valid = true;
    set<char> seen;

    for (char c : P) {
        if (seen.count(c)) {
            is_valid = false;
            break;
        }
        seen.insert(c);
        if (S.find(c) == string::npos) {
            is_valid = false;
            break;
        }
    }

    long long count = 0;

    if (is_valid) {
        int S_length = S.length();
        int P_length = P.length();
        int D = S_length - P_length;

        count = 1;

        int index_range = 1;

        for (int i = 2; i <= D + 1; ++i) {
            long long temp = 1;
            for (int j = 0; j < index_range; ++j) {
                temp *= static_cast<long long>(D - j);
            }
            count += temp * i;
            index_range++;
        }
    }

    cout << count << '\n';
    return 0;
}

File: 백준/Silver/24371. БРОЙ ПОДНИЗОВЕ/БРОЙ ПОДНИЗОВЕ.py
S = input()
P = input()

is_valid = all([1 if P.count(p) == 1 and p in S else 0 for p in P])
count = 0

if is_valid:
    S_length = len(S)
    P_length = len(P)
    D = S_length - P_length

    count += 1

    multiply_list = [d for d in range(D, 0, -1)]
    index_range = 1

    for i in range(2, D + 2):
        temp = 1
        
        for j in range(index_range):
            temp *= multiply_list[j]

        count += temp * i
        index_range += 1

print(count)